doc/PROTOCOL-SEMANTICS.md

```diff
@@ -0,0 +1,268 @@
+normally handled by the Channel.
+
+A Method's intended operation is just normal documentation describing what a
+Method does, intended for a developer. gRPC itself is not generally aware of the
+intended operation.
+
+A Method has a request message type and a separate response message type. gRPC
+is only aware of these types well enough to serialize and deserialize them.
+Messages are considered opaque byte sequences of a known length to gRPC itself.
+
+A Method's request and response each have a cardinality: either one ("unary"),
+or zero to many (a "stream"). This produces four possible configurations which,
+for convenience, each have a name:
+
+|                       | **unary response** | **streaming response** |
+| --------------------- | ------------------ | ---------------------- |
+| **unary request**     | unary              | server-streaming       |
+| **streaming request** | client-streaming   | bidirectional (bidi)   |
+
+While we use the term "method," pedantically it is closer to "function" as it is
+not object oriented and there is no "receiver" involved (the `this` variable in
+many languages) other than the destination machine. gRPC is based on message
+passing, not object orientation.
+
+Related Methods are typically grouped into a Service. To gRPC, a Service is a
+group of methods that tend to be implemented together and that all share the
+Service's namespace. A Service is a higher-level abstraction and may not be
+present explicitly in all implementations. However, the namespace provided by a
+Service is a core distinguishing feature of its Methods; if two Methods have the
+same name but exist in different Services they must be considered distinct and
+not be confused. A Method name including its Service namespace prefix with a "/"
+separator is a "full method name".
+
+## Calls
+
+RPCs, or "Calls," are initiated by a client to a server, typically via a
+Channel. There may be multiple servers that _could_ have received the Call (as
+is common for load balancing), but only a single server may process an
+individual Call. Calls are assumed to be non-idempotent and may not be
+"replayed" except for when gRPC is explicitly informed it is safe to do so.
+
+Calls are natively two independent streams (i.e., full duplex bidirectional) of
+Messages. The request stream is started with Request Headers and ended by Half
+Close. The response stream is started with Response Headers and ended by
+Trailers, or consistes only of Trailers. Messages may exist between the headers
+and the end of the stream. Request Headers, Response Headers, Messages, Half
+Close, and Trailers are the units of communication and, absent the Call's
+termination, will be communicated to the remote without the need to send further
+units on the stream. However, see the optimizations permitted for unary Calls
+below.
+
+Request Headers contain the Full Method Name and Metadata. Response Headers
```

Comment:
Request Headers contains?  (Throughout this paragraph you treat Headers/Trailers as plural, but elsewhere you treat them as singular.)

I would actually prefer that we call these "Header" and "Trailer" to avoid this problem, since you can't have more than one Header/Trailer on a stream (per direction).  I know we use plural when we talk about these, probably for historical reasons, but it seems most correct to refer to them as singular.